src/QsCompiler/DataStructures/SyntaxTree.fs

```diff
@@ -671,6 +673,37 @@ and QsStatement =
     }
 
 
+/// The source files for a syntax tree node.
+type Source =
+    {
+        /// The path to the original source code file.
+        CodeFile: string
+
+        /// The path to the assembly file if the node was loaded from a reference.
+        AssemblyFile: string QsNullable
+    }
+
+    /// The assembly file path for this source if one exists, otherwise the code file path.
+    member source.AssemblyOrCode = source.AssemblyFile.ValueOr source.CodeFile
+
+    /// <summary>
+    /// Returns a copy of this source with the given <paramref name="codeFile"/> or <paramref name="assemblyFile"/> if
+    /// provided.
+    /// </summary>
+    member source.With([<Optional; DefaultParameterValue null>] ?codeFile,
+                       [<Optional; DefaultParameterValue null>] ?assemblyFile) =
+        { source with
+            CodeFile = codeFile |> Option.defaultValue source.CodeFile
+            AssemblyFile = assemblyFile |> QsNullable<_>.FromOption |> QsNullable.orElse source.AssemblyFile
+        }
+
+/// Operations for source files.
+module Source =
```

Comment:
I don't have a ton of experience with F# so maybe this question is a bit naive. Why is a `Source` module needed? When an object of `Source` type is present, wouldn't be enough just to call the `AssemblyOrCode` member of the object?

Reply:
It's not strictly necessary, but it's more convenient and idiomatic to use module functions instead of properties in F#. Functions support type inference:

```fsharp
let foo source =
    "File: " + Source.assemblyOrCode source

let foo (source: Source) = // Type annotation is needed.
    "File: " + source.AssemblyOrCode
```

Functions are also easier to pass as a value than a property getter:

```fsharp
let paths = sources |> List.map Source.assemblyOrCode

// Lambda is needed.
let paths = sources |> List.map (fun source -> source.AssemblyOrCode)
```